### ISSUE TITLE
Possible bug in HBIOS DIOMEDIA function

### ISSUE THREAD

**Issue by hubertushirsch** [wwarthen/RomWBW#415]
Hi Wayne,

I have either found a bug in RomWBW or have misunderstood something.

I have built a SC126 computer (Stephen Cousins, Z180). It has an SD card adapter, equipped with a 4GB SD card.
The SC126 was delivered with RomWBW 3.4.0, I updated to 3.5.0-dev.71.
I formatted the SD card in hd1k format with space for 32 slices in the RomWBW partition.
```
FDISK80 for RomWBW, UNA, Mini-M68k, KISS-68030, SBC-188  ----
       Version 1.1-23 created 3-June-2023
                 (Running under RomWBW HBIOS)

HBIOS unit number [0..4]: 4
Capacity of disk 4:  (  4G)  7744512      Geom 762c1010
Nr  ---Type- A --      Start         End   LBA start  LBA count  Size
 1    RomWBW   2e      8:0:1  1023:15:16        2048     524288  256M
 2             00       *** empty ***
 3             00       *** empty ***
 4             00       *** empty ***
```
The CP/M drives correctly display 1024 directory entries.
```
B>stat c:dsk:

    C: Drive Characteristics
65408: 128 Byte Record Capacity
 8176: Kilobyte Drive  Capacity
 1024: 32  Byte Directory Entries
    0: Checked  Directory Entries
  256: Records/ Extent
   32: Records/ Block
   64: Sectors/ Track
    2: Reserved Tracks
```
I did some tests with HBIOS calls under the RomWBW monitor.
I don't understand the return parameter of the DIOMEDIA function.
```
        ld      bc,18h*100h+4   ; DIOMEDIA Unit 4
        ld      e,1             ; Perform discovery; e=0 w/o discovery
        rst     8h
        call    printstat
results in
A=00 BC=00ff DE=0004 HL=5693    ; if called with E=1 (Perform discovery)
A=00 BC=1804 DE=0004 HL=001d    ; if called with E=0 (w/o discovery)
```
In any case, E=4 is returned, which according to the table in the RomWBW System Guide means "MID_HD512 4 Hard Disk (LBA) w/ 512 directory entries".
I expected the return value E=10 (0xa) at this point.
"MID_HD1K 10 Hard Disk (LBA) w/ 1024 directory entries"
Is this a bug or is this intended?

Then another note:
In the RomWBW System Guide 9.4.12 it should perhaps be pointed out that the return value in D is the (number_of_heads | 0x80) for LBA drives. That means the correct number of heads is obtained by (D & 0x7f).
```
        ld      bc,1bh*100h+4   ; DIOGEOMETRY Unit 4
        rst     8h
        call    printstat
results in
A=00 BC=0200 DE=9010 HL=762c    ; D=0x10 | LBA_Bit
```
Greetings
Hubert

**Comment by wwarthen** [comment]
Hi @hubertushirsch,

It is actually operating as intended, but I understand that it is not obvious what is going on and I did not document it well.

I am out of town until Monday and I want to fully explain what is happening.  I will post a complete response on Monday.

Thank you, Wayne

**Comment by wwarthen** [comment]
Hi @hubertushirsch,

OK, I have a bit more time now, so let me expand on the problem you have encountered.

With respect to the output of the DIOMEDIA function, as you observed, the function is returning with E=4 (MID_HD512) even when the disk device is partitioned as MID_HD1K.  The problem occurs because the HBIOS does not know how a hard disk is partitioned.  By design, it does not examine the partition table.  So, it will always return MID_HD512 for all hard disk media.

The differentiation between MID_HD_512 and MID_HD1K is made by the operating systems.  When the operating systems see a hard disk (MID_512), they then read and interpret the partition table of the hard disk device to further determine if the format is MID_HD512 or MID_HD1K.

I fully appreciate this is confusing and the current HBIOS documentation does not properly describe this.  I will leave this issue open until I have updated the documentation.

At this point, you are proobably wondering why there is an MID_HD1K enumaration, if it is never returned by the HBIOS.  It is because it is used internally by the OSes.  It is not really correct that the MID_HD1K enumaration exist in the HBIOS interface and I am going to take a look at removing it.

Regarding the DIOGEOMETRY function, you are also correct that I need to document the use of the high bit of D (which indicates the device is LBA capable).  I will do that as well.

Again, this issue will remain open until I have corrected the documentation.  I am extremely backed up right now with some commitments I need to deal with first.  So, it may take me a few days to get the documentation corrected.

Thank you!

Wayne

**Comment by hubertushirsch** [comment]
Hi Wayne

Thanks for the explanation. From my point of view this (HBIOS) is an abstraction layer for different hardware variants.

I don't think it's helpful to rewrite the additional code for every application and every OS that a user wants to implement again and again. HD512 and HD1K are two precisely documented variants of the hard drive. The return of the recognized variant by the HBIOS would be a good basis to support the user in writing consistent programs.

Let me write about my thoughts in more detail tonight.

Greetings
Hubert

**Comment by hubertushirsch** [comment]
Hi Wayne

I have now looked at the source code and seen that the call to the DIOMEDIA function is forwarded to the respective driver.
So several drivers would be affected by the change (sd.asm, ide.asm, hdsk.asm and others). That would be a lot of work.
Another aspect is that existing operating systems or standalone programs would have to be changed, as they do not currently expect E=10 as a result (e.g. CBIOS).
So I accept the situation as it currently is.

I also noticed that in the System Guide (3.5.0) the media IDs are named differently than in hbios.inc.
MID_HD512 => MID_HD
MID_HD1K  => MID_HDNEW

Regards
Hubert
